Conversation Patterns Docs (#6373)
RasaHQ/rasa: docs/docs/chitchat-faqs.mdx

  Bot: My name is Sara!
User: Which languages can I build assistants in?
  Bot: You can use Rasa to build assistants in any language you want!
User: What’s Rasa X?
  Bot: Rasa X is a tool to learn from real conversations and improve your assistant.
```

## Step-by-step Guide on Using Retrieval Actions for FAQs and Chitchat

To handle FAQs and chitchat you'll need a rule-based dialogue management policy ([the `RulePolicy`](#rulepolicy)) and an easy way
to return the appropriate response for a question ([the `ResponseSelector`](#retrieval-actions-and-responseselector)).

### 1. Updating the configuration

For FAQs and chitchat, you always want the assistant to respond the same way every time
the same type of question is asked. [Rules](rules.mdx) allow you to do exactly that.
To use rules, the you need to add the [`RulePolicy`](#rule-policy) to your policies in your configuration file:

```rasa-yaml
policies:
# other policies
- name: RulePolicy
```

The `RulePolicy` works alongside other policies in your configuration (like the `TEDPolicy`). Its job is to enforce rule-based logic for conversation paths where you already know how the assistant should respond to a given input.

> To use retrieval actions, include the `ResponseSelector` in your NLU pipeline in your configuration file.

Author: ```suggestion
To use retrieval actions, include the ResponseSelector in your NLU pipeline in your configuration file.
```

> The `ResponseSelector` requires a featurizer and intent classifier to work, so

Author: ```suggestion
The ResponseSelector requires a featurizer and intent classifier to work, so
```

id: chitchat-faqs
sidebar_label: Chitchat and FAQs
title: Chitchat and FAQs
abstract: FAQ assistants are the simplest assistants to build and typically the first kind of assistant anyone builds. This page is a guide to the concepts and training data you need to handle non-contextual questions like FAQs and chitchat.
---

[FAQs](./glossary.mdx#faqs) and [chitchat](./glossary.mdx#chitchat) are two cases where the conversational assistant's answer is a single message,

> and the assistant should always answer the same way,

Reviewer: hm, it's not really a requirement that the assistants answer is a single message, it can also be multiple. the only requirement is that the user only sends one message.

no matter what has happened previously in the conversation.
For example, in the
following conversation, every question's answer is independent of context and
could

> just as easily be answered on its own:

Reviewer: is there a reason this is split over so many lines? Also, can we rephrase this to "For example, in the following conversation, every question can be asked at any point in the conversation and the answer is independent of anything the user has said previously" or similar

> ```yaml

Reviewer: there will be a chat component available soon which lu is working on, we should use that when it's ready

User: What's your name
  Bot: My name is Sara!
User: Which languages can I build assistants in?
  Bot: You can use Rasa to build assistants in any language you want!
User: What’s Rasa X?
  Bot: Rasa X is a tool to learn from real conversations and improve your assistant.
```

> ## Using Retrieval Actions for FAQs and Chitchat

Reviewer: I would once again say "Step-by-step guide" here as in the other pages

> To handle FAQs and chitchat you'll need a rule-based dialogue management policy ([the `RulePolicy`](#rulepolicy)) and an easy way

Reviewer: ```suggestion
To handle FAQs and chitchat you'll need a rule-based dialogue management policy (the [RulePolicy](#rulepolicy)) and an easy way
```

> to return the appropriate response for a question ([the `ResponseSelector`](#retrieval-actions-and-responseselector)).

Reviewer: ```suggestion
to return the appropriate response for a question (the [ResponseSelector](#retrieval-actions-and-responseselector)).
```

### 1. Updating the configuration

> For FAQs and chitchat, you always want the same response (or action) for the same question.

Reviewer: ```suggestion
For FAQs and chitchat, you always want the assistant to respond the same way every time the same type of question is asked.
```

[Rules](rules.mdx) allow you to do exactly that.

> To use rules, the you need to add the [`RulePolicy`](#rule-policy) to your policies in your configuration file:

Reviewer: ```suggestion
To use rules, the you need to add the [RulePolicy](#rule-policy) to your policies in your configuration file:
```

```rasa-yaml
policies:
# other policies
- name: RulePolicy
```

> The `RulePolicy` works alongside other policies in your configuration (like the `TEDPolicy`). Its job is to enforce rule-based logic for conversation paths where you already know how the assistant should respond to a given input.

Reviewer: hm, not sure i would say that "rule-based logic for conversation paths where you already know how the assistant should respond to a given input" - arguably you always know what the assistant should respond.
Other question - do we think we need to explain how the rule policy functions here? (not necessarily saying we shouldn't, just asking for your opinion

Author: From what we talked about re. keeping reasons for using and not explanations, no, I think we don't need to

### 2. Defining Retrieval Actions and the `ResponseSelector`

> Consider an example where you have 20 different FAQs. Although each question is represented as an individual intent, all FAQ intents are handled the same way in the dialogue. For each FAQ intent, the assistant **retrieves** the proper response depending on which question has been asked.

Reviewer: "retrieves" doesn't really mean anything to the user. Is there a reason for the lengthy explanation here and below?

we had a very short explanation originally:
```
The [ResponseSelector](components/selectors.mdx#responseselector) is designed to
make it easier to handle conversation patterns like small talk and FAQ messages.
When you use the `ResponseSelector`, you only need one story to handle all FAQs,
instead of adding one story for each intent.
```
^ probably shouldn't be exactly that, but i don't think we need to do the whole "instead of writing 20 rules" thing

Instead of writing 20 rules, you can use [retrieval
actions](retrieval-actions.mdx) to handle all FAQs with a single rule by grouping them together under a single [retrieval intent](glossary.mdx#retrieval-intent)
called e.g. `faq`.

The retrieval action uses the output of the
[`ResponseSelector`](./components/selectors.mdx#response-selector) to return

> the correct response for the specific FAQ that the user asked.

Reviewer: again not sure we need to explain this concept?

To use retrieval actions, include the `ResponseSelector` in your NLU pipeline in your configuration file.

> The `ResponseSelector` requires a featurizer and intent classifier to work, so

Reviewer: let's not ` format these.

title: Chitchat and FAQs
---

FAQs and chitchat both involve non-contextual questions with single-message answers. 
For example, in the following conversation, every question's answer is independent of context and could
just as easily be answered on its own:

> User: What's your name
>
>> Bot: My name is Sara!
>
> User: Which languages can I build assistants in?
> 
>> Bot: You can use Rasa to build assistants in any language you want!
> 
> User: What’s Rasa X?
>
>> Bot: Rasa X is a tool to learn from real conversations and improve your assistant. Read more [here](https://rasa.com/docs/rasa-x/)

This page is a guide on the concepts and training data you need to handle non-contextual questions like
FAQ and chitchat.

> You can use the example configuration and training data on this page to build an assistant

Reviewer 2: This was a bit confusing, because the reader isn't likely to want to exactly replicate the example convo? Maybe it would be better to say, 

The example described here can be adapted to other use cases where your assistant needs to return the right answer to single-turn questions and interactions.

Author: good point, was not sure how tutorial-y to keep it. I realize there's some missing pieces if someone was trying to follow it copy-paste style, but I don't think we quite want to prepare for that. I'll adapt your phrasing and maybe change the end section too.

that can handle a conversation like the one above, 
or you can apply the guidelines to your own use case.

## Concepts

> To handle FAQ and chitchat you'll need a rule-based dialogue management policy ([the `RulePolicy`](#rulepolicy)), and an easy way

Reviewer 2: also nit-picky :) no comma after RulePolicy

I also see a few places where we use FAQ when it should be plural (FAQs)

to return the appropriate response for a question ([the `ResponseSelector`](#retrieval-actions-and-responseselector)).  

### `RulePolicy`

For FAQ and chitchat, you always want the same response (or action) for the same question. 
[Rules](rules.mdx) allow you to do exactly that.
To use rules, the you need to add the [`RulePolicy`](#rule-policy) to your policies in your configuration file:

```yaml
policies:
- name: RulePolicy
```

> You can have other policies as well, but the `RulePolicy` is the only one necessary for handling rules.

Reviewer 2: "can have other policies" seems a bit vague. Maybe a bit more context here?

The RulePolicy works alongside other policies in your configuration (like the TED Policy). Its job is to enforce rule-based logic for conversation paths where you already know how the assistant should respond to a given input.

### Retrieval Actions and `ResponseSelector`

> With FAQ and chitchat, you often have multiple intents that should be handled

Reviewer 2: I wonder if it might help to reword slightly. Something like,

Consider an example where you have 20 different FAQs. Although each question is represented as an individual intent, all FAQ intents are handled quite similarly in the dialogue. For each FAQ intent, the assistant retrieves the proper response depending on which question has been asked. Instead of writing 20 rules, retrieval actions allow you to handle this logic with a single rule.

title: Chitchat and FAQs
---

> FAQs and chitchat are two cases where the conversational assistant's answer is

Author: No glossary entry, added short explanation

    epochs: 100
```

By default, the `ResponseSelector` will build a single retrieval model for all retrieval intents.

> If you want to retrieve responses for FAQs and chitchat separately, you can use multiple `ResponseSelector` components

Author: ```suggestion
If you want to retrieve responses for FAQs and chitchat separately, you can use multiple ResponseSelector components
```

and specify the `retrieval_intent` key, for example:

```rasa-yaml
pipeline:
# Other components
- name: ResponseSelector
  epochs: 100
  retrieval_intent: faq
- name: ResponseSelector
  epochs: 100
  retrieval_intent: chitchat
```

> It's recommended to start with separate models for each `retrieval_intent`. You can merge them later if you find it works better for your use case.

Author: ```suggestion
```

### 2. Defining Retrieval Actions and the `ResponseSelector`

Consider an example where you have 20 different FAQs. Although each question is represented as an individual intent, all FAQ intents are handled the same way in the dialogue. For each FAQ intent, the assistant **retrieves** the proper response depending on which question has been asked.

Instead of writing 20 rules, you can use [retrieval
actions](retrieval-actions.mdx) to handle all FAQs with a single rule by grouping them together under a single [retrieval intent](glossary.mdx#retrieval-intent)
called e.g. `faq`.

The retrieval action uses the output of the
[`ResponseSelector`](./components.mdx#response-selector) to return
the correct response for the specific FAQ that the user asked.

### 3. Creating rules

> When using retrieval actions, you need to write only one rule for each retrieval intent. All intents

Author: ```suggestion
You need to write only one rule for each retrieval intent. All intents
```

    epochs: 100
```

By default, the `ResponseSelector` will build a single retrieval model for all retrieval intents.
If you want to retrieve responses for FAQs and chitchat separately, you can use multiple `ResponseSelector` components

> and specify the `retrieval_intent` key, for example:

Reviewer: should this be phrased more like a recommendation? instead of saying "if you want to"

```rasa-yaml
pipeline:
# Other components
- name: ResponseSelector
  epochs: 100
  retrieval_intent: faq
- name: ResponseSelector
  epochs: 100

>   retrieval_intent: chitchat

Reviewer: this should probably be in the "updating the configuration section"

```

> It's recommended to start with separate models for each `retrieval_intent`. You can merge them later if you find it works better for your use case.

Reviewer: "it's recommended" is a no-no phrase according to our style guide :D i think we can get rid of this sentence if the section before the pipeline definition is phrased as a recommendation

### 3. Creating rules

> When using retrieval actions, you need to write only one rule for each retrieval intent. All intents

Reviewer: this may be starting to get a little too technical? Maybe we can do this more from a "group of FAQs" perspective rather than saying "when using retrieval actions" all the time

grouped under that retrieval intent will then be handled the same way.
The retrieval action starts with `respond_` and ends with the retrieval intent's name.

> For example, rules for responding to FAQs and chitchat would look like this:

Reviewer: i think we should stop saying "for example". If the idea was to have this page act as a full tutorial and make you have a functioning assistant by the end then we should tell the user what to do, i.e. "write these rules"

    - action: respond_chitchat
```

The actions `respond_faq` and `respond_chitchat` will use the `ResponseSelector`'s prediction to return the actual response message.

### 4. Updating the NLU Training Data

NLU training examples for the `ResponseSelector` look the same as
regular training examples, except that their names must refer to the retrieval
intent they are grouped under. For example:

```rasa-yaml
nlu:
  - intent: faq/ask_channels
    examples: |
      - What channels of communication does rasa support?
      - what channels do you support?
      - what chat channels does rasa uses
      - channels supported by Rasa
      - which messaging channels does rasa support?
  - intent: faq/ask_languages
    examples: |
      - what language does rasa support?
      - which language do you support?
      - which languages supports rasa
      - can I use rasa also for another language?
      - languages supported
  - intent: faq/ask_rasax
    examples: |
      - I want information about rasa x
      - i want to learn more about Rasa X
      - what is rasa x?
      - Can you tell me about rasa x?
      - Tell me about rasa x
      - tell me what is rasa x
```

> Be sure to update your domain file to include the added `faq` intent.

Reviewer: how?

### 5. Defining the responses

Responses for the `ResponseSelector` follow the same naming convention as
retrieval intents. Besides this, they can have all the characteristics of
normal bot [response](domain.mdx#responses). For the FAQs and chitchat intents
listed above, our responses could look like:

```rasa-yaml
responses:
  faq/ask_channels:
  - text: |
    We have a comprehensive list of [supported connectors](https://rasa.com/docs/core/connectors/), but if
    you don't see the one you're looking for, you can always create a custom connector by following
    [this guide](https://rasa.com/docs/rasa/user-guide/connectors/custom-connectors/).
  faq/ask_languages:
  - text: |
    You can use Rasa to build assistants in any language you want!
  faq/ask_rasax:
  - text: |
    Rasa X is a tool to learn from real conversations and improve your assistant.

>     Read more [here](https://rasa.com/docs/rasa-x/)

Reviewer: weren't we going to use the examples from the `responseselectorbot` here btw?

```

## Summary

Once you've done the following, you can train your bot and try it out!

- [ ] Add `RulePolicy` to your policies in `config.yml`

> - [ ] Add `ResponseSelector` to your pipeline in `config.yml`

Reviewer: these two should be one step

    epochs: 100
```

> By default, the `ResponseSelector` will build a single retrieval model for all retrieval intents.

Reviewer 2: When might you want to build separate models for different intents vs a single model for all retrieval intents?

Author: I'm actually second guessing myself on this now - @akelad what would you say?

Reviewer: Given that we train on intent labels now it prob makes sense to just do separate models each time. but @dakshvar22 do you have opinions on that?

If you want to retrieve responses for FAQ and chitchat seperately, you can use multiple `ResponseSelector` components 
and specify the `retrieval_intent` key, for example:

```yaml
pipeline:
# Other components
- name: ResponseSelector
  epochs: 100
  retrieval_intent: faq
- name: ResponseSelector
  epochs: 100
  retrieval_intent: chitchat
```

## Training Data

To train a Rasa model using the `RulePolicy` and the `ResponseSelector`, you'll need to provide
rules, NLU training examples, and responses.

### Rules

> By using retrieval actions, you only need to write one rule for each retrieval intent. All intents

Reviewer 2: By using retrieval actions => When using retrieval actions

normal bot [response](domain.mdx#responses). For the FAQ and chitchat intents
listed above, our responses could look like:

```yaml
responses:
  faq/ask_channels: 
  - text: |
    We have a comprehensive list of [supported connectors](https://rasa.com/docs/core/connectors/), but if
    you don't see the one you're looking for, you can always create a custom connector by following
    [this guide](https://rasa.com/docs/rasa/user-guide/connectors/custom-connectors/).
  faq/ask_languages:
  - text: |
    You can use Rasa to build assistants in any language you want!
  faq/ask_rasax:
  - text: |
    Rasa X is a tool to learn from real conversations and improve your assistant. 
    Read more [here](https://rasa.com/docs/rasa-x/)
```

## Putting it all together

Once you've done the following, you can train your bot and try it out!

- [ ] Added `RulePolicy` to your policies in `config.yml`
- [ ] Added `ResponseSelector` to your pipeline in `config.yml`
- [ ] Added at least one rule for responding to FAQ/chitchat
- [ ] Added examples for your FAQ/chitchat intents
- [ ] Added responses for your FAQ/chitchat intents

> Even if you have other skills in your bot besides chitchat and FAQs, your bot should be able to respond consistently 

Reviewer 2: `Even if you have other skills in your bot` feels a bit clunky. Maybe,

Now, your assistant can respond correctly and consistently to FAQs and chitchat, even if these interjections happen while your assistant is helping the user with another task.
